src/main/java/com/microsoft/aad/msal4j/ServerSideTelemetry.java

```diff
@@ -0,0 +1,116 @@
+    }
+
+    static void addFailedRequestTelemetry(String publicApiId, String correlationId, String error){
+
+        String[] previousRequest = new String[]{publicApiId, error};
+        previousRequests.put(
+                correlationId,
+                previousRequest);
+    }
+
+    static void incrementSilentSuccessfulCount(){
+        silentSuccessfulCount.incrementAndGet();
+    }
+
+    synchronized CurrentRequest getCurrentRequest() {
+        return currentRequest;
+    }
+
+    synchronized void setCurrentRequest(CurrentRequest currentRequest) {
+        this.currentRequest = currentRequest;
+    }
+
+    private synchronized String buildCurrentRequestHeader(){
+        if(currentRequest == null){
+            return StringHelper.EMPTY_STRING;
+        }
+
+        return SCHEMA_VERSION +
+                SCHEMA_PIPE_DELIMITER +
+                currentRequest.publicApi().getApiId() +
+                SCHEMA_COMMA_DELIMITER +
+                currentRequest.forceRefresh() +
+                SCHEMA_PIPE_DELIMITER;
+    }
+
+    private synchronized String buildLastRequestHeader() {
+
+        // LastRequest header schema:
+        // schema_version|silent_successful_count|api_id1,correlation_id1|error1|
+        StringBuilder lastRequestBuilder = new StringBuilder();
+
+        lastRequestBuilder
+                .append(SCHEMA_VERSION)
+                .append(SCHEMA_PIPE_DELIMITER)
+                .append(silentSuccessfulCount.getAndSet(0));
+
+        if (previousRequests.isEmpty()) {
+            // Kusto queries always expect all delimiters so return
+            // "schema_version|silent_successful_count|||"
+            return lastRequestBuilder
+                    .append(SCHEMA_PIPE_DELIMITER)
+                    .append(SCHEMA_PIPE_DELIMITER)
+                    .append(SCHEMA_PIPE_DELIMITER)
+                    .toString();
+        }
+
+        StringBuilder middleSegmentBuilder = new StringBuilder(SCHEMA_PIPE_DELIMITER);
+        StringBuilder errorSegmentBuilder = new StringBuilder(SCHEMA_PIPE_DELIMITER);
+
+        Iterator<String> it = previousRequests.keySet().iterator();
+
+        // Total header size should be less than 8kb. At max, we will use 4kb for telemetry.
+        while (it.hasNext()
+                && (middleSegmentBuilder.length() + errorSegmentBuilder.length()) < 3800)
+        {
+            String correlationId = it.next();
+            String[] previousRequest = previousRequests.get(correlationId);
+            String apiId = (String)Array.get(previousRequest, 0);
+            String error = (String)Array.get(previousRequest, 1);
+
+            middleSegmentBuilder.append(apiId).append(SCHEMA_COMMA_DELIMITER).append(correlationId);
+            errorSegmentBuilder.append(error);
+
+            if(it.hasNext()){
+                middleSegmentBuilder.append(SCHEMA_COMMA_DELIMITER);
+                errorSegmentBuilder.append(SCHEMA_COMMA_DELIMITER);
+            }
+
+            it.remove();
```

Comment:
Seems like you are deleting the requests one at a time, would it make sense to only clear after a the request is successfully sent? 
also what happens if there are 2 requests going on a the same time? Will the first one here get all the elements and the next request will be waiting until this is complete?

Reply:
Good point. I have updated to keep track of requests in progress. In case of HTTP status code 429 or >500, requests are not deleted. 

If there are 2 requests going on at the same time, for the same client application object, the first one will get all of the elements (up to 4kb of data), and flush them. The second thread will be blocked. (only while the request header is built). If there is any data left, it will flush that. If not, then it would send a blank previous request header.